File: leetcode/linked-list/876. Middle of the Linked List/py/main.py
class Solution:    
    # Time O(N)
    # Space O(1)
    def middleNode(self, head: ListNode) -> ListNode:
        if head == None or head.next == None:
            return head

        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
          
        return slow

File: leetcode/linked-list/876. Middle of the Linked List/js/main.js
// Time O(N)
// Space O(1)
const middleNode = head => {
  if (head == null || head.next == null) {
    return head;
  }

  let slow = head;
  let fast = head;

  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
  }
  return slow;
};
